### PR TITLE
[cling] Startup script support

### PR DESCRIPTION
**Pull request by jiangyilism**
## Changes or fixes:

Implement startup script support according to this post:
https://root-forum.cern.ch/t/cling-startup-scripts-support/51358

## Checklist:

- [X] tested changes locally
- [ ] updated the docs (if necessary)

### CODE REVIEW

**Comment by jiangyilism** [comment]
cling startup scripts are different from rootlogon.C and rootrc .
rootrc is a config file while .clingrc is a regular cling script. cling startup scripts are executed also in invocation of standalone cling binary (not from root interpreter).

Question:
1. Renaming `.cling_profile` and `.clingrc` with suffix `.C` ?
2. Rename  `.clingrc` to something else so that `.clingrc` can be reserved for cling config file in the future (if any)?
Making it a config file instead of a cling script aligns with rootrc but not with bashrc, zshrc.
3. Drop `.clingrc`  for now and keep `.cling_profile` only ?  If a cling script can easily tell if it is in interactive mode then keeping only 1 script makes sense. Otherwise it is better to keep both to align with bash and other interpreter inspired by bash design. By easily telling I mean a macro like `CLING_INTERACTIVE` or something simple that does not access `gCling`.

**Comment by vgvassilev** [comment]
@phsft-bot build!

@jiangyilism, thanks for you PR, can you add some tests for the new functionality under `cling/test`?

**Comment by jiangyilism** [comment]
Hi @vgvassilev, `cling/test/StratupFile.C` is added. However it only covers usage of `.cling_profile.C`.

To proceed further, I would like to clerify the above decisions first.

Let me summarize again here:

1. Should we .x execute the startup file(s) or execute it in shebang `#!` way?
   @eguiraud, @vgvassilev, @Axel-Naumann, what is the designed/documented difference between the 2?
   Will it effect project root (which I have not tried out myself. I am experimenting standalone cling only)
   As far as I read from the web, `rootlogon.C` and `.rootrc` are executed in shebang way.
   Their filename-functions `rootlogon()`, `rootrc()` are not executed.

   My proposal: Execute startup files in .x way 

2. How should the startup files be named?   `.cling_profile`, `.cling_login`, `.cling_logon` like bash and root?

   https://www.gnu.org/software/bash/manual/html_node/Bash-Startup-Files.html

   Should we adopt file suffix `.C` ?

   My proposal: `.cling_profile.C` to align with rootlogon.C  (to be honest I just want the editor syntax highlighting...)

3. Do we need `.cling_profile` and `.clingrc` as the original draft implementation?
   If inside the script we can tell if cling is in interactive mode by a macro like `CLING_INTERACTIVE` then
   only 1 file is needed.

   My proposal: keep `.cling_profile.C` as the only startup file. Forget about `#define CLING_INTERACTIVE 1` for now.

   I mentioned this support just because bash and other shells/interpreters have it.
   But unless someone wants this feature right now, I prefer not to implement it and
   reserve `CLING_INTERACTIVE` and `.clingrc`/`.clingrc.C` for future design.

**Comment by jiangyilism** [comment]
@Axel-Naumann , @vgvassilev Could you provide some feedback?

**Comment by vgvassilev** [comment]
@jiangyilism, thanks for your patch. I am a bit worried about the fact that upon startup cling will execute "random" code. Can you elaborate a little more about your usecase?

I suspect python does something similar, can we model the feature following their approach?

**Comment by jiangyilism** [comment]
My use case is to `#include <...>` some standard headers I often use. e.g. filesystem and thread.
I will also put `namespace fs = std::filesystem;` in my startup file.
cling already implicitly includes some standard headers so I think letting users to customize it makes sense.

About "execute random code", is there risk if the executed code/script is in user's home directory or `XDG_CONFIG_HOME` ?

**Comment by vgvassilev** [comment]
> My use case is to `#include <...>` some standard headers I often use. e.g. filesystem and thread. I will also put `namespace fs = std::filesystem;` in my startup file. cling already implicitly includes some standard headers so I think letting users to customize it makes sense.

Thanks for explaining. It all makes sense.

> About "execute random code", is there risk if the executed code/script is in user's home directory or `XDG_CONFIG_HOME` ?

Well, I guess bash and python suffer from the same issue. As long as we are consistent to their practices it should be no issue.

**Comment by vgvassilev** [comment]
> cling startup scripts are different from rootlogon.C and rootrc . rootrc is a config file while .clingrc is a regular cling script. cling startup scripts are executed also in invocation of standalone cling binary (not from root interpreter).
> 
> Question:
> 
>     1. Renaming `.cling_profile` and `.clingrc` with suffix `.C` ?
> 
>     2. Rename  `.clingrc` to something else so that `.clingrc` can be reserved for cling config file in the future (if any)?
>        Making it a config file instead of a cling script aligns with rootrc but not with bashrc, zshrc.
> 
>     3. Drop `.clingrc`  for now and keep `.cling_profile` only ?  If a cling script can easily tell if it is in interactive mode then keeping only 1 script makes sense. Otherwise it is better to keep both to align with bash and other interpreter inspired by bash design. By easily telling I mean a macro like `CLING_INTERACTIVE` or something simple that does not access `gCling`.

How about a `~/.cling.d` folder where we glob all files and execute?

**Comment by jiangyilism** [comment]
> How about a `~/.cling.d` folder where we glob all files and execute?
 
Sounds good to me too.

I will enumerate `.C` files in `.cling.d/` (if dir exists) with `llvm::vfs::directory_iterator` like [CompilerInstance.cpp](https://clang.llvm.org/doxygen/CompilerInstance_8cpp_source.html#l00248). So the enumeration does not guarantee specific orderings.

Search order of `.cling.d/` is still
1. ${CLING_HOME} envvar
2. ${XDG_CONFIG_HOME}/cling/
3. ${HOME}/.config/cling/
4. ${HOME}/

Does that look good to you?

**Comment by Axel-Naumann** [comment]
Apologies for not reacting - this sounds terrific! Could you update the PR?

**Comment by jiangyilism** [comment]
> So the enumeration does not guarantee specific orderings.

Updated. However the behavior is slightly different: the startup files are sorted after enumeration so the execution order is deterministic.

File: interpreter/cling/tools/driver/cling.cpp [hunk 1]
```diff
@@ -15,14 +15,17 @@
...
 #include "llvm/Support/ManagedStatic.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/PrettyStackTrace.h"
+#include "llvm/Support/Signals.h"
 
-#include <iostream>
+#include <algorithm>
 #include <fstream>
-#include <vector>
+#include <iostream>
```
Comment (Axel-Naumann) [inline interpreter/cling/tools/driver/cling.cpp:26]: Thanks!

**Comment by Axel-Naumann** [comment]
@phsft-bot build

File: interpreter/cling/tools/driver/cling.cpp [hunk 2]
```diff
@@ -51,6 +54,58 @@ static int checkDiagErrors(clang::CompilerInstance* CI, unsigned* OutErrs = 0) {
...
+  }
+
+  if (llvm::sys::path::home_directory(DirPath)) {
+    return DirPath;
+  }
+
+  return {};
+}
+
+static void runStarupFiles(cling::UserInterface& Ui) {
```
[the remark this answers is not in the document]
Comment (jiangyilism) [reply interpreter/cling/tools/driver/cling.cpp]: Fixed. Thanks.

**Comment by vgvassilev** [comment]
@jiangyilism I was wondering what's the status of this PR?

**Comment by jiangyilism** [comment]
> @jiangyilism I was wondering what's the status of this PR?

There is no pending issue as far as I know. The mac11/cxx14 test fail above is due to (flaky?) timeout so it seems irrelevant to this PR.

**Comment by vgvassilev** [comment]
@Axel-Naumann, should we merge this then?

**Comment by Axel-Naumann** [comment]
Thanks for your very nice contribution!

**Pull request merged**